# Meteor Challenge

Para este desafio eu usei a linguagem Python utilizando a biblioteca Python Image Library (PIL)

In [1]:
from PIL import Image

Inicialmente, eu carreguei a imagem utilizando a biblioteca de python PIL e, de acordo com o enunciado, criei as variáveis que representam os pixels relevantes da imagem.

In [2]:
image = Image.open('meteor_challenge_01.png')

#Converte os pixels imagem para o modo rgb
image = image.convert('RGB')

star_pixel = (255,255,255)
meteor_pixel = (255,0,0)
water_pixel = (0,0,255)
ground_pixel = (0,0,0)


## Quantidade de estrelas

Para descobrir a quantidade de estrelas, percorro o vetor de pixels dado pela função ```getdata()``` e conto a quantidade de pixels brancos encontrados.

In [3]:
star_count = 0
for pixel in image.getdata():
   if (pixel == star_pixel):
       star_count += 1
   elif (pixel == ground_pixel):
       break

print("Number of star is: ",star_count) 

Number of star is:  315


## Quantidade de meteoros
Da mesma forma que feito anteriormente, só que agora é encontrar a quantidade de pixels vermelhos.

In [4]:
meteor_count = 0
for pixel in image.getdata():
    if (pixel == meteor_pixel):
        meteor_count += 1
    elif (pixel == ground_pixel):
        break

print("Number of meteor is: ",meteor_count) 

Number of meteor is:  328


## Quantidade de meteoros que cairão na água
Neste caso, irei transformar a imagem em uma matriz de pixels. Dessa forma percorrerei coluna por coluna da imagem contando a quantidade meteoros, caso no final eu encontro o pixel azul (que representa a água), a quantidade meteoros encontradas daquela coluna irá para a solução final.

In [5]:
# armazena a altura e largura da figura selecionada
width = image.size[0]
height = image.size[1]

#carrega a matriz de pixels da imagem
pixels = image.load()

meteor_water_count = 0

#percorre as colunas da matriz, caso encontre agua no final da coluna armazena o valor obtido
for i in range(width):
    aux = 0 
    for j in range(height):
        if (pixels[i,j] == meteor_pixel):
            aux += 1
        elif (pixels[i,j] == water_pixel):
            meteor_water_count += aux
            break

print("Number of meteors that fell into the water: ",meteor_water_count)

Number of meteors that fell into the water:  105


## Frase escondida
Nesta etapa opcional, eu tentei algumas coisas mas não obtive sucesso. Provavelmente, deve ser uma ideia muito mais simples das que eu tive.

-------------------------------
A primeira coisa que tentei, foi tentar mudar o fundo do sol para ver se a mensagem estava ali. Entretanto, nada foi encontrado. A imagem ```teste.png``` mostra o resultado.

In [6]:
image = Image.open("meteor_challenge_01.png")
image.convert("RGB")
pixels = image.load()
width = image.size[0]
height = image.size[1]

for i in range(width):
    for j in range(height):
        if (pixels[i,j] == ground_pixel):
            pixels[i,j] = star_pixel
            
image.save("test.png")

Minha segunda tentativa foi tentar aplicar técnica de stenografia para descobrir a mensagem. No caso usei o método do Least Significant Bit. Este método consiste em pegar o bit menos signifcante de cada pixel (que no caso possui 8 bits), e a cada 8 pixels um novo byte surgia.

In [7]:
from PIL import Image

#change array to char
def arr_to_char(arr):
    expoent = 0
    value = 0
    for bit in arr:
        value += int(bit)*2**expoent
        expoent+=1
    return chr(value)

image = Image.open("meteor_challenge_01.png")
image = image.convert("L")

data = ''

count = 0
arr_bytes = []
arr = []
for pixel in image.getdata():
    arr.append(bin(pixel)[-1])
    count += 1
    if(count % 8 == 0):
        arr_bytes.append(arr_to_char(arr))
        arr = []

string = ''
for char in arr_bytes:        
    if (char != '\x00'):
        string += char
        
print(string)



" 		€
€" 		$€€" 		€€" 		
"0ˆFF뀆~€âÀ:@q@|£ƒ1üc€¸žÐ?¢4à€(àÑAþô ÿÝðŠ
À£ƒûàYààì{px4@p$0‡Ì!


-------------------------------------
# Conclusão

Estou enviando junto o script com os algoritmos criados de forma resumida.

| Task                          | Answer|
|------------------------------	|-----	|
| Number of stars              	| 315 	|
| Number of Meteors            	| 328 	|
| Meteors falling on the Water 	| 105 	|
| (optional) Hidden Phrase     	|     	|